Asyncronously render composer map previews

nyalldawson opened the pull request:
New PR for the background async rendering of composer map previews

The version in https://github.com/qgis/QGIS/pull/4494 caused some crashes on Travis . Will rework here to fix these.

nyalldawson committed:
[FEATURE] Asyncronously render composer map previews without blocking UI

Makes working with compositions much more responsive and enjoyable!

Inspired by manisandro's commit in Sourcepole's fork

nyalldawson committed:
Rework how composer map cache is invalidated

Avoids unnecessary redraws and defers cache update until
when item is actually visible

nyalldawson committed:
sipify QgsComposerMap

nyalldawson committed:
[composer] Always update map caches when zoom changes

Before this would only happen with the "render" mode, not "cache"
mode. Now updating caches is so responsive and transparent to users
we can safely trigger the background cache update whenever the
zoom changes without any performance impact.

nyalldawson committed:
[needs-docs][composer] Remove Rectangle mode from composer maps

No longer required - the Cache mode is now performant enough
that we don't have to rely on the old rectangle preview
when desiring responsive composers

nyalldawson commented:
@nirvn Care to test this reworked branch?

nyalldawson committed:
If composer map is set to use project CRS, update cached render
when project CRS changes

This should prevent one of the most common confusions with
composer maps - users inserting multiple maps, with apparently
different CRS due to outdated cache renders, and getting confused
because outputs from composer look different to the composer view....

nirvn commented:
@nyalldawson , it works really well over here.  The cache deference until a map is visible makes a significant difference too (on my older machine).

nirvn commented:
So, basically, I've decided to stick to this branch until it's merged onto master. I do _not_ want to go back! 😉

nyalldawson committed:
[needs-docs][composer] Remove Render/Cache modes from composer maps

Make maps invalidate cache whenever layer order changes. Cache mode
is now identical to Render mode, so  all mode switching has been
removed from composer maps. This should greatly simplify the learning
curve and usability of composer maps.
